C++: Add default value handling to mutation/SetField code

llchan opened the pull request:
If a field set to the default value is mutated to the same default value, mutation returns a failure even though it can succeed by leaving the buffer unchanged. This change passes the default value to SetField so that it can be smarter about accepting these types of mutations.

I'm not very familiar with the reflection code but I did my best to cover my bases there as well. Some extra eyes on that would be good.

Addresses issue #4084

llchan commented:
I signed it!

aardappel commented on include/flatbuffers/reflection.h:
```diff
@@ -243,7 +255,28 @@ template<typename T> T *GetAnyFieldAddressOf(const Struct &st,
   assert(sizeof(T) == GetTypeSize(field.type()->base_type()));
-  return table->SetField(field.offset(), val);
+  T def;
+  switch (field.type()->base_type()) {
```
this may be quicker with an if?
if (type <= ULong) ..
else if (type <= Double) ..
else..
?

llchan replied:
That would work yeah, but are we worried about that being brittle if the enum values change? If that's something we don't care about, I'm okay with either implementation. Imho the switch makes the intent more clear, but nbd either way.

llchan replied:
Nevermind, just realized those enum values are generated from the reflection.fbs, and if we're following the normal schema upgrade conventions those shouldnt be changed.

I'll update PR.

llchan commented:
Update: I saw that flatbuffers.h had `IsScalar`/`IsInteger`/`IsFloat` helpers, so I added the equivalents for the reflection types, and I updated the code to use them.

aardappel commented on include/flatbuffers/reflection.h:
```diff
@@ -242,8 +263,20 @@ template<typename T> T *GetAnyFieldAddressOf(const Struct &st,
+  T def;
+  if (IsInteger(type)) {
+    def = GetFieldDefaultI<T>(field);
+  } else if (IsFloat(type)) {
```
You've already checked that it is scalar above, so this can lose the is `IsFloat` check (or become `.. else { assert(IsFloat(..`

llchan replied:
I put that in there in case we ever add any scalars that aren't ints/floats, but if we're good with the the assumption that types arent going to be added without changes here I can turn it into an assert.

aardappel commented:
Looks good, one small nit.

aardappel commented:
Looks great, thanks!
